### PR TITLE
Keep the replica count same for kcm if the shoot is not transitioning from/to hibernated state

### LINKED ISSUE #1866: Kube-controller-manager replica count is 0 when cluster is un-hibernated
**What happened**:
When a hibernated cluster is woken up, the kube-controller-manager deployment's replica count remains 0. Because of this the nodes are not able to join.

**What you expected to happen**:
The kube-controller-manager replica count should be 1 for a running cluster

**How to reproduce it (as minimally and precisely as possible)**:
When dependency watchdog is not running, wake up a hibernated cluster.

**Anything else we need to know?**:
This is probably because of these lines:
https://github.com/gardener/gardener/blob/v0.35.2/pkg/operation/botanist/controlplane.go#L999-L1003
```golang
	replicaCount, err := common.CurrentReplicaCount(b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeControllerManager)
	if err != nil {
		return err
	}
	defaultValues["replicas"] = replicaCount
```
There should be a `if b.Shoot.HibernationEnabled` check for this.

**Environment**:

- Gardener version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #1866

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

**Comment by rfranzke** [comment]
See issue. I don't understand when this is happening. We have https://github.com/gardener/gardener/blob/9e15bb19e29de2e45e4a29685835fc172759db64/pkg/operation/botanist/controlplane.go#L215-L252 that explicitly wakes KCM up. Can you provide the steps to reproduce this?

**Comment by ggaurav10** [comment]
This seems to be called only when deleting a hibernated cluster. Let me know if i understood this wrong. The issue is for hibernated clusters that is just woken up and not getting deleted.

**Comment by rfranzke** [comment]
Thanks. Still, steps for reproduction would be helpful to understand the problem.

**Comment by ggaurav10** [comment]
Preliminarily it looks like it has something to do with dependency watchdog. Needs some more digging.
i can see the following error messages in dependency-watchdog logs regarding the failing cluster:
```
I0128 03:10:31.634774       1 prober.go:243] shoot-kube-apiserver/shoot--mlfprod--preprod-eu/internal: probe failed with error: the server has asked for the client to provide credentials. Will retry...
```

Nevertheless, Gardener seems to be dependent on dependency watchdog to scale up the kube-controller-manager when waking up a cluster, and doesn't scale up the deployment itself. If dependency watchdog is down, the kcm pod is never deployed, but the next in line deployments (eg MCM) are scaled up by Gardener itself. IMHO, Gardener should be able to atleast create/wakeup a cluster on its own without any dependency on the dependency-watchdog.
wdyt?

**Comment by ggaurav10** [comment]
Preliminarily it looks like it has something to do with dependency watchdog. Needs some more digging.
i can see the following error messages in dependency-watchdog logs regarding the failing cluster:
```
I0128 03:10:31.634774       1 prober.go:243] shoot-kube-apiserver/shoot--<redacted>--<redacted>/internal: probe failed with error: the server has asked for the client to provide credentials. Will retry...
```

Nevertheless, Gardener seems to be dependent on dependency watchdog to scale up the kube-controller-manager when waking up a cluster, and doesn't scale up the deployment itself. If dependency watchdog is down, the kcm pod is never deployed, but the next in line deployments (eg MCM, which starts creating machine objects which fail to join cluster) are scaled up by Gardener itself. IMHO, Gardener should be able to atleast create/wakeup a cluster on its own without any dependency on the dependency-watchdog.
wdyt?

**Comment by rfranzke** [comment]
@ggaurav10 makes sense, the shoot reconciliation loop should not depend on DWD for scale-up. +1

**Comment by rfranzke** [comment]
Though, with your PR, what I understand is that only when `.spec.hibernation.enabled=true` the replica count for KCM will not be changed, correct? I think @amshuman-kr's motivation was that even if `.spec.hibernation.enabled=false` the shoot reconciliation should NOT accidentally scale-up KCM again if DWD has scaled it down. How do we solve this?

**Comment by ggaurav10** [comment]
Can we rely on `.status.IsHibernated` and `.spec.hibernation.enabled` together to figure out that the shoot is waking up or hibernating, and based on that take a decision whether to keep the replica count same or change?

**Comment by rfranzke** [comment]
Yes, this would work. Thank you. Please adapt accordingly.

File: pkg/operation/botanist/controlplane.go [hunk 1]
```diff
@@ -996,11 +996,15 @@ func (b *Botanist) DeployKubeControllerManager() error {
 		},
 	}
 
-	replicaCount, err := common.CurrentReplicaCount(b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeControllerManager)
-	if err != nil {
-		return err
+	if b.Shoot.HibernationEnabled == b.Shoot.Info.Status.IsHibernated {
+		// Keep the replica count same if the shoot is not trasitioning from/to hibernation state
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
		// Keep the replica count same if the shoot is not transitioning from/to hibernation state
```

**Comment by amshuman-kr** [comment]
> Can we rely on .status.IsHibernated and .spec.hibernation.enabled together to figure out that the shoot is waking up or hibernating, and based on that take a decision whether to keep the replica count same or change?

Yes. That sounds better.

**Review by rfranzke — COMMENTED** [review]
looks good now, just a typo in the comment

**Review by rfranzke — APPROVED** [review]

**Review by timebertt — APPROVED** [review]
/lgtm
Actually, if stumbled upon these lines somewhen in the last few days, when the kcm of a dev Shoot was not scaled up properly.
I can't remember the exact scenario, but I think, this PR should fix the issue.

File: pkg/operation/botanist/controlplane.go [hunk 2]
```diff
@@ -996,11 +996,15 @@ func (b *Botanist) DeployKubeControllerManager() error {
...
-		return err
+	if b.Shoot.HibernationEnabled == b.Shoot.Info.Status.IsHibernated {
+		// Keep the replica count same if the shoot is not transitioning from/to hibernation state
+		// otherwise this may interfere with dependency-watchdog
+		replicaCount, err := common.CurrentReplicaCount(b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeControllerManager)
+		if err != nil {
+			return err
+		}
+		defaultValues["replicas"] = replicaCount
 	}
```
Comment (danielfoehrKn) [inline pkg/operation/botanist/controlplane.go:1007]: The defaultValues["replicas"] from the chart contains 1 - so this is now actually scaling up the KCM if the Shoot is not hibernated.  Can we make this clear in a comment somewhere for code better readability?

Comment (rfranzke) [reply pkg/operation/botanist/controlplane.go:1007]: WDYM @danielfoehrKn ? If the shoot is not hibernated then it should be deployed with replicas=1, similar to every other component we deploy. Why do we need a comment for this? Or am I overseeing something?

**Review by amshuman-kr — APPROVED** [review]
LGTM. Thanks @ggaurav10  for finding the bug as well as the right solution approach.

File: pkg/operation/botanist/controlplane.go (hunk 2 above)
Comment (rfranzke) [reply pkg/operation/botanist/controlplane.go:1007]: OK, as discussed, we should not use any Helm values as defaults. Instead, we should always explicitly inject the value as we need them for the various charts. The `values.yaml` file should only be an example showing which values are possible in general.
We should straighten this soon. For now, let's continue with this PR.

**Pull request merged**